### PR TITLE
Wang-Buzsaki model added.

### PR DESCRIPTION
**Pull request by Ziaeemehr**
Wang-Buzsaki model, conductance model synapse corresponding to 
_Gamma Oscillation by Synaptic Inhibition in a Hippocampal Interneuronal Network Model_,  Journal of neuroscience 1996.

fig_cond

```python
import nest
import pylab as pl

# nest.Install("testModuleInstallmodule")
nest.Install("nestml_module")
model = "wb_cond_exp_nestml"

dt = 0.01
tfinal = 100.0
nest.SetKernelStatus({
    "resolution": dt})

neuron = nest.Create(model)
parameters = nest.GetDefaults(model)

if 0:
    for i in parameters:
        print(i, parameters[i])


nest.SetStatus(neuron, {'I_e': 75.0})
multimeter = nest.Create("multimeter")
nest.SetStatus(multimeter, {"withtime": True,
                            "record_from": ["V_m"],
                            "interval": dt})
spikedetector = nest.Create("spike_detector",
                            params={"withgid": True,
                                    "withtime": True})
nest.Connect(multimeter, neuron)
nest.Connect(neuron, spikedetector)
nest.Simulate(tfinal)
dmm = nest.GetStatus(multimeter)[0]
Voltages = dmm["events"]["V_m"]
tv = dmm["events"]["times"]
dSD = nest.GetStatus(spikedetector, keys='events')[0]
Spikes = dSD['senders']
ts = dSD["times"]

fig, ax = pl.subplots(2, figsize=(8, 6), sharex=True)
ax[0].plot(tv, Voltages, lw=2, color="k")
ax[1].plot(ts, Spikes, 'ko')
ax[1].set_xlabel("Time [ms]")
ax[1].set_xlim(0, tfinal)
ax[1].set_ylabel("Spikes")
ax[0].set_ylabel("v [ms]")
ax[0].set_ylim(-100, 50)

for i in ts:
    ax[0].axvline(x=i, lw=1., ls="--", color="gray")

pl.savefig("fig_cond.png")
pl.show()
```

### CODE REVIEW

**Comment by Ziaeemehr** [comment]
I added the receptors Ampa, NMDA, Gabba_a and Gabba_b receptors to Wang-Buzsaki model.
The receptors corresponts to the model `hill_tononi`.
fig_recep


```python
import nest
import pylab as pl

nest.Install("nestml_module")
model = "wb_receptors_nestml"

dt = 0.01
tfinal = 100.0
nest.SetKernelStatus({
    "resolution": dt})

neuron = nest.Create(model)
parameters = nest.GetDefaults(model)

if 0:
    for i in parameters:
        print(i, parameters[i])

nest.SetStatus(neuron, {'I_e': 75.0})
multimeter = nest.Create("multimeter")
nest.SetStatus(multimeter, {"withtime": True,
                            "record_from": ["V_m"],
                            "interval": dt})
spikedetector = nest.Create("spike_detector",
                            params={"withgid": True,
                                    "withtime": True})
nest.Connect(multimeter, neuron)
nest.Connect(neuron, spikedetector)
nest.Simulate(tfinal)
dmm = nest.GetStatus(multimeter)[0]
Voltages = dmm["events"]["V_m"]
tv = dmm["events"]["times"]
dSD = nest.GetStatus(spikedetector, keys='events')[0]
Spikes = dSD['senders']
ts = dSD["times"]

fig, ax = pl.subplots(2, figsize=(8, 6), sharex=True)
ax[0].plot(tv, Voltages, lw=2, color="k")
ax[1].plot(ts, Spikes, 'ko')
ax[1].set_xlabel("Time [ms]")
ax[1].set_xlim(0, tfinal)
ax[1].set_ylabel("Spikes")
ax[0].set_ylabel("v [ms]")
ax[0].set_ylim(-100, 50)

for i in ts:
    ax[0].axvline(x=i, lw=1., ls="--", color="gray")

pl.savefig("fig_recep.png")
pl.show()

```

**Comment by Ziaeemehr** [comment]
Reduced Traub-Miles Model of a Pyramidal Neuron in Rat Hippocampus according to _An Introduction to Modeling Neuronal Dynamics_ chapter 5 and Traub and Miles, Neuronal Networks of the Hippocampus 1991 added.


```python
import nest
import pylab as pl

nest.Install("nestml_module")
model = "traub_psc_alpha_nestml"

dt = 0.01
tfinal = 100.0
nest.SetKernelStatus({
    "resolution": dt})

neuron = nest.Create(model)
parameters = nest.GetDefaults(model)

nest.SetStatus(neuron, {'I_e': 150.0})
multimeter = nest.Create("multimeter")
nest.SetStatus(multimeter, {"withtime": True,
                            "record_from": ["V_m"],
                            "interval": dt})
spikedetector = nest.Create("spike_detector",
                            params={"withgid": True,
                                    "withtime": True})
nest.Connect(multimeter, neuron)
nest.Connect(neuron, spikedetector)
nest.Simulate(tfinal)
dmm = nest.GetStatus(multimeter)[0]
Voltages = dmm["events"]["V_m"]
tv = dmm["events"]["times"]
dSD = nest.GetStatus(spikedetector, keys='events')[0]
Spikes = dSD['senders']
ts = dSD["times"]
```
fig
coupled

**Comment by Ziaeemehr** [comment]
Reduced Traub-Miles Model of a Pyramidal Neuron in Rat Hippocampus according to An Introduction to Modeling Neuronal Dynamics chapter 5 and Traub and Miles, Neuronal Networks of the Hippocampus, 1991 added. The receptors are according to `hill_tononi` model.

**Example:**

Coupled Traub neurons by Ampa receptors.


ampa_coupled

```python
import nest
import pylab as pl
import matplotlib

matplotlib.rc('font', size=14)
nest.Install("nestml_module")
model = "traub_receptors_nestml"

nest.set_verbosity("M_WARNING")
nest.ResetKernel()


dt = 0.01
tfinal = 1000.0
nest.SetKernelStatus({
    "resolution": dt})

neuron1 = nest.Create(model, 1)
nest.SetStatus(neuron1, {'I_e': 100.0})

neuron2 = nest.Create(model)
nest.SetStatus(neuron2, {"AMPA_Tau_1": 0.1,
                         "AMPA_Tau_2": 2.4,
                         "AMPA_g_peak": 0.1,
                         })


multimeter = nest.Create("multimeter", 2)
nest.SetStatus([multimeter[0]], {"withtime": True,
                                 "record_from": ["V_m"],
                                 "interval": dt})


record_from = ["V_m", "I_syn_ampa", "I_syn_nmda", "I_syn_gaba_a", "I_syn_gaba_b"]

nest.SetStatus([multimeter[1]], {"withtime": True,
                                 "record_from": record_from,
                                 "interval": dt})
#! {'AMPA': 1, 'NMDA': 2, 'GABA_A': 3, 'GABA_B': 4}
# nest.Connect(neuron1, neuron2)
nest.Connect(neuron1, neuron2, syn_spec={"receptor_type": 1})  # AMPA
nest.Connect(neuron1, neuron2, syn_spec={"receptor_type": 2})  # NMDA
nest.Connect(neuron1, neuron2, syn_spec={"receptor_type": 3})  # GABAA
nest.Connect(neuron1, neuron2, syn_spec={"receptor_type": 4})  # GABAB

nest.Connect([multimeter[0]], neuron1, "one_to_one")
nest.Connect([multimeter[1]], neuron2)

spikedetector = nest.Create("spike_detector",
                            params={"withgid": True,
                                    "withtime": True})
nest.Connect(neuron1, spikedetector)
nest.Simulate(tfinal)

fig, ax = pl.subplots(3, figsize=(8, 6), sharex=True)

dmm = nest.GetStatus(multimeter)[1]
Voltages = dmm["events"]["V_m"]
tv = dmm["events"]["times"]

ax[0].plot(tv, Voltages, lw=2, label=str(2))

labels = ["ampa", "nmda", "gaba_a", "gaba_b"]
j = 0
for i in record_from[1:]:
    g = dmm["events"][i]
    ax[1].plot(tv, g, lw=2, label=labels[j])
    j +=1

dSD = nest.GetStatus(spikedetector, keys='events')[0]
Spikes = dSD['senders']
ts = dSD["times"]

ax[2].plot(ts, Spikes, 'k.')
ax[2].set_xlabel("Time [ms]")
ax[2].set_xlim(0, tfinal)
ax[2].set_ylabel("Spikes")

ax[0].set_title("recording from PSP")
ax[0].set_ylabel("v [ms]")
# ax[0].set_ylim(-68, -66)
ax[1].set_ylabel("I_syn")
ax[1].legend(frameon=False, loc="upper right")

pl.savefig("ampa_coupled.png")
# pl.show()
```

**Comment by clinssen** [comment]
Hi, thanks very much for these contributions! It would be great if we could also somehow merge the example scripts that you provided. To this end, could we ask you to turn them into simple unit tests? This way we can also validate properties of the model.

One check that you could use, is to measaure the average firing rate (total spikes fired divided by the simulation interval), and check that it is close to some desired value (e.g. 50 spikes/s). Alternatively you could do the same for the inter-spike-interval.

Use e.g. https://github.com/nest/nestml/blob/master/tests/nest_tests/nest_multisynapse_test.py as a template, and before plotting, add an assertion such as "assert abs(actual_rate - expected_rate) < failure_threshold", to make the test fail in case the results are off.

Thanks in advance!

File: models/traub_receptors.nestml [hunk 1]
```diff
@@ -0,0 +1,202 @@
+"""
+Name: traub_psc_alpha - Traub model according to Borgers 2017.
```
Comment (clinssen) [inline models/traub_receptors.nestml:2]: What do you think of renaming to "traub_cond_multisyn",  to fit the existing naming scheme better?

File: models/wb_receptors.nestml [hunk 2]
```diff
@@ -0,0 +1,198 @@
+"""
+Name: wb_receptors - Wang buzsaki model with
```
Comment (clinssen) [inline models/wb_receptors.nestml:2]: Rename to "wb_cond_multisyn"?

File: models/traub_psc_alpha.nestml [hunk 3]
```diff
@@ -0,0 +1,117 @@
+"""
+Name: traub_psc_alpha - Traub model according to Borgers 2017.
+
+Reduced Traub-Miles Model of a Pyramidal Neuron in Rat Hippocampus[1].
+parameters got from reference [2].
+
+(1) Post-syaptic currents
```
Comment (clinssen) [inline models/traub_psc_alpha.nestml:7]:
```suggestion
(1) Post-synaptic currents
```

File: models/traub_receptors.nestml (hunk 1 above)
Comment (Ziaeemehr) [reply models/traub_receptors.nestml:2]: I agree with renaming, why not. I work on the unit tests. 
Thank you.

**Comment by clinssen** [comment]
@Ziaeemehr: for legal reasons, we ask contributors to sign a copyright transfer agreement. Your contributions will be part of the next NESTML release. We use Zenodo to generate a citable record of the release with DOI (see e.g. https://zenodo.org/record/3697733 for the current release). You will be fully credited there as a co-author on the next release (if you want!)

I would like to hereby kindly ask @heplesser to get in touch with you with the agreement form.

**Comment by Ziaeemehr** [comment]
> @Ziaeemehr: for legal reasons, we ask contributors to sign a copyright transfer agreement.  You will be fully credited there as a co-author on the next release (if you want!)
> 
> I would like to hereby kindly ask @heplesser to get in touch with you with the agreement form.
 
@clinssen: That's fine. 
Please let me know if I need to modify the unit tests.
regards,

**Comment by Ziaeemehr** [comment]
> @Ziaeemehr: for legal reasons, we ask contributors to sign a copyright transfer agreement.  You will be fully credited there as a co-author on the next release (if you want!)
> 
> I would like to hereby kindly ask @heplesser to get in touch with you with the agreement form.
 
@clinssen: That's fine. 
Please let me know if I need to modify the unit tests.
regards,

**Comment by Ziaeemehr** [comment]
@clinssen  How can I do the checks myself to see the errors and debug. I run the tests on my machine without error.

**Comment by clinssen** [comment]
In the test scripts, the path to the NEST installation (`nest_path`) has to be updated from "/home/abolfazl/prog/install-dir/nest-simulator-2.18.0_build" to the right Travis path: "/home/travis/nest_install"

**Comment by clinssen** [comment]
@Ziaeemehr: I have filed a pull request against your branch that should fix the unit tests. Could you please review and merge the PR? Thanks!

**Comment by Ziaeemehr** [comment]
@clinssen  Thanks for your changes.

**Pull request merged**